src/lua/protobuf.lua

```diff
@@ -0,0 +1,642 @@
+local ffi = require('ffi')
+local wireformat = require('internal.protobuf.wireformat')
+local protocol_mt
+-- These constants are used to define the boundaries of valid field ids.
+-- Described in more detail here:
+-- https://protobuf.dev/programming-guides/proto3/#assigning
+local MIN_FIELD_ID = 1
+local RESERVED_FIELD_ID_MIN = 19000
+local RESERVED_FIELD_ID_MAX = 19999
+local MAX_FIELD_ID = 2^29 - 1
+
+-- Number limits for int32 and int64
+local MAX_FLOAT = 0x1.fffffep+127
+local MAX_UINT32 = 2^32 - 1
+local MAX_UINT64 = 0xfffffffffffff800
+local MIN_SINT32 = -2^31
+local MAX_SINT32 = 2^31 - 1
+local MIN_SINT64 = -0x8000000000000000
+local MAX_SINT64 = 0x7ffffffffffffc00
```

Comment:
Added necessary comments

Reply:
Quote:

> ```lua
> -- Actual uint64 limit is 2^64 - 1. Because of lua number limited precision
> -- numbers from [2^64 - 1024, 2^64 + 2048] represent as 2^64. So the correct
> -- number limit for uint64 is 2^64 - 1025.
> local MAX_UINT64 = 0xfffffffffffff800
> ```

However, 0xfffffffffffff800 is 2^64 - 2048. 2^64 - 1025 is 0xfffffffffffffbff.

Quote:

> ```lua
> -- Same problem with lua number limited precision.
> -- Numbers from [2^63 - 512, 2^63 + 1024] represent as 2^63. So the correct
> -- number limit for int64 is 2^63 - 513.
> local MAX_INT64 = 0x7ffffffffffffc00
> ```

0x7ffffffffffffc00 is 2^63 - 1024. 2^63 - 513 is 0x7ffffffffffffdff.

Maybe determine the numbers dynamically? A kind of the following:

```lua
local function first_less_then(x)
    local d = 1
    while x - d == x do
        d = d + 1
    end
    return x - d
end

local MAX_UINT64 = first_less_then(2^64)
local MAX_INT64 = first_less_then(2^63)
```

Quote:

> ```lua
> local MIN_INT64 = -0x8000000000000000
> ```

We agreed to add `-- 2^63` comment, but I see no such a comment.

Quote:

> ```lua
> -- Sint64 values are encoded using 'ZigZag' algorithm, so negative values
> -- are encoded as positive, which leads to the same problem with upper
> -- limit for int64.
> local MAX_SINT64 = 0x7ffffffffffffc00
> local MIN_SINT64 = -0x7ffffffffffffc00
> ```

TBH, I don't understand why a destination form (zigzag, 2's complements) matters here.

Also, it is not clear why minimum int64 and minimum sint64 values are different. Maybe I'm missing something here.

Reply:
It's strange but I checked and got this, I think it may be rounded again. I will fix this to be correct
```
tarantool> 0xfffffffffffffbff == 0xfffffffffffff800
---
- true
...
```

Same issue:
```
tarantool> 0x7ffffffffffffc00 == 0x7ffffffffffffdff
---
- true
...
```

Yes, I forgot to return this comment after I changed SINT limits

In ZigZag for negative values we calculate positive value to encode by formula `2 * (-value) - 1`. If we take `MIN_INT64` we will get resulting positive value out of range because it will be 2^64 because of rounding.

Reply:
Still, how about something of the `first_less_then` kind?

----

> > TBH, I don't understand why a destination form (zigzag, 2's complements) matters here.
> >
> > Also, it is not clear why minimum int64 and minimum sint64 values are different. Maybe I'm missing something here.

> In ZigZag for negative values we calculate positive value to encode by formula `2 * (-value) - 1`. If we take `MIN_INT64` we will get resulting positive value out of range because it will be 2^64 because of rounding.

Am I understand right that it is a restriction of the current implementation? I think that we can use `cdata<int64_t>` or `cdata<uint64_t>` as a storage for temporary values to perform the zig-zag encoding transformations even if the input value type is `number`. I guess that such an implementation wouldn't add extra restrictions.

Reply:
I checked limits and replaced them with right ones. I don't know about `first_less_then` because these values are one time calculated and I hope they won't change, so I think this function will be excess.

Yes, this is a restriction of current implementation because calculations are made in `number` values and not `cdata`.  I tested it using `cdata` cast before performing calculations and it will work in `cdata<uint64_t>` case, so I will implement this.

Reply:
> I checked limits and replaced them with right ones. I don't know about `first_less_then` because these values are one time calculated and I hope they won't change, so I think this function will be excess.

It has the same effect, adds negligible startup time calculations and more readable for a human. The readability seems important for me.

Anyway, I don't mind constant values. It also has advantages: at least, predicatability and less area for bugs like cycling on startup and similar ones.

Let's proceed with the given solution.

----

> Yes, this is a restriction of current implementation because calculations are made in `number` values and not `cdata`.  I tested it using `cdata` cast before performing calculations and it will work in `cdata<uint64_t>` case, so I will implement this.

Nice! Thanks.